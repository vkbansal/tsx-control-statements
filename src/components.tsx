import React from 'react';

export interface IfProps {
  condition: boolean;
  children: React.ReactNode;
}

export function If(props: IfProps): React.ReactElement | null {
  return props.condition ? <React.Fragment>{props.children}</React.Fragment> : null;
}

If.__ControlStatement = Symbol('If');

export interface ForProps<T> {
  items: T[];
  children: (item: T, index: number) => React.ReactNode;
}

export function For<T>(props: ForProps<T>): React.ReactElement | null {
  return <React.Fragment>{props.items.map(props.children)}</React.Fragment>;
}

For.__ControlStatement = Symbol('For');

export type SingleOrArray<T> = T | T[];

export interface ChooseProps {
  children: React.ReactNode;
}

export function Choose(props: ChooseProps): React.ReactElement | null {
  const children = React.Children.toArray(props.children);

  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (
      React.isValidElement<WhenProps>(child) &&
      (child.type as typeof When).__ControlStatement === When.__ControlStatement &&
      child.props.condition
    ) {
      return <React.Fragment>{child}</React.Fragment>;
    }

    if (
      React.isValidElement<OtherWiseProps>(child) &&
      (child.type as typeof Otherwise).__ControlStatement === Otherwise.__ControlStatement
    ) {
      return <React.Fragment>{child}</React.Fragment>;
    }
  }

  return null;
}

Choose.__ControlStatement = Symbol('Choose');

export interface WhenProps {
  condition: boolean;
  children: React.ReactNode;
}

export function When(props: WhenProps): React.ReactElement | null {
  return props.condition ? <React.Fragment>{props.children}</React.Fragment> : null;
}

When.__ControlStatement = Symbol('When');

export interface OtherWiseProps {
  children: React.ReactNode;
}

export function Otherwise(props: OtherWiseProps): React.ReactElement | null {
  return <React.Fragment>{props.children}</React.Fragment>;
}

Otherwise.__ControlStatement = Symbol('Otherwise');
