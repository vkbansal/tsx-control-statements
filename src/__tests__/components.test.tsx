import React from 'react';
import { render } from '@testing-library/react';
import { expect, describe, test } from 'vitest';

import { If, For, Choose, When, Otherwise } from '../components';

describe('component tests', () => {
	describe('<If /> tests ', () => {
		// @vitest-environment happy-dom
		test('works with truthy condition', () => {
			const { container } = render(<If condition={true}>I&apos;m inside a truthy condition</If>);

			expect(container).toMatchInlineSnapshot(`
        <div>
          I'm inside a truthy condition
        </div>
      `);
		});

		// @vitest-environment happy-dom
		test('works with falsy condition', () => {
			const { container } = render(<If condition={false}>I&apos;m inside a falsy condition</If>);

			expect(container).toMatchInlineSnapshot(`<div />`);
		});
	});

	describe('<For /> tests ', () => {
		// @vitest-environment happy-dom
		test('works with arrow function', () => {
			const { container } = render(
				<For items={[1, 2, 3, 4, 5]}>{(item) => <div key={item}>{item}</div>}</For>,
			);

			expect(container).toMatchInlineSnapshot(`
        <div>
          <div>
            1
          </div>
          <div>
            2
          </div>
          <div>
            3
          </div>
          <div>
            4
          </div>
          <div>
            5
          </div>
        </div>
      `);
		});

		// @vitest-environment happy-dom
		test('works with normal function', () => {
			const { container } = render(
				<For items={[1, 2, 3, 4, 5]}>
					{function (item) {
						return <div key={item}>{item}</div>;
					}}
				</For>,
			);

			expect(container).toMatchInlineSnapshot(`
        <div>
          <div>
            1
          </div>
          <div>
            2
          </div>
          <div>
            3
          </div>
          <div>
            4
          </div>
          <div>
            5
          </div>
        </div>
      `);
		});
	});

	describe('<Choose /> tests ', () => {
		// @vitest-environment happy-dom
		test('works with only <When /> and <OtherWise />', () => {
			const { container } = render(<Choose>Hello {'world'.toUpperCase()} !</Choose>);

			expect(container).toMatchInlineSnapshot(`<div />`);
		});

		// @vitest-environment happy-dom
		test('works with <When />', () => {
			const { container } = render(
				<Choose>
					<When condition={false}>Hello {'world'.toLowerCase()} 1 !</When>
					<When condition={true}>Hello {'world'.toUpperCase()} !</When>
					<When condition={false}>Hello {'world'.toLowerCase()} 2 !</When>
				</Choose>,
			);

			expect(container).toMatchInlineSnapshot(`
				<div>
				  Hello 
				  WORLD
				   !
				</div>
			`);
		});

		// @vitest-environment happy-dom
		test('works with <Otherwise />', () => {
			const { container } = render(
				<Choose>
					<When condition={false}>Hello {'world'.toLowerCase()} 1 !</When>
					<When condition={false}>Hello {'world'.toLowerCase()} 2 !</When>
					<Otherwise>Hello {'world'.toUpperCase()} !</Otherwise>
				</Choose>,
			);

			expect(container).toMatchInlineSnapshot(`
				<div>
				  Hello 
				  WORLD
				   !
				</div>
			`);
		});
	});
});
