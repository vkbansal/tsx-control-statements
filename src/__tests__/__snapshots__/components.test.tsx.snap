// Vitest Snapshot v1

exports[`component tests > <Choose /> tests  > works with <Otherwise /> 1`] = `
<div>
  Hello 
  WORLD
   !
</div>
`;

exports[`component tests > <Choose /> tests  > works with <When /> 1`] = `
<div>
  Hello 
  WORLD
   !
</div>
`;

exports[`component tests > <For /> tests  > works with arrow function 1`] = `
<div>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
  <div>
    4
  </div>
  <div>
    5
  </div>
</div>
`;

exports[`component tests > <For /> tests  > works with normal function 1`] = `
<div>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
  <div>
    4
  </div>
  <div>
    5
  </div>
</div>
`;

exports[`component tests > <If /> tests  > works with truthy condition 1`] = `
<div>
  I'm inside a truthy condition
</div>
`;
